### PR TITLE
CDAP-13268 Ability for the user to consolidate field operations at Workflow level.

### PR DESCRIPTION
**Pull request by sagarkapare**
This PR includes MR changes. Will add spark as separate commit. Opening on top of #10078 since this PR uses ability to publish to TMS from it.

### CODE REVIEW

File: cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java [hunk 1]
```diff
@@ -66,4 +67,14 @@
...
+
+  /**
+   * Call this method to disable emitting the field lineage operations from nodes running inside Workflow.
+   * Disabling should happen from {@link AbstractWorkflow#initialize} method at which point no node
+   * has been executed yet. Calling this method means Workflow is taking responsibility of emitting
+   * the field operation. Field operations will be available to workflow as {@link WorkflowNodeState}
+   * by calling {@link #getNodeStates} method.If workflow does not call {@link LineageRecorder#record}
+   * method, then no field lineage will be recorded.
+   */
+  void disableFieldLineageOperationsEmitFromNodes();
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java:79]: hmmm that is a long method name. How about ```enableFieldLineageConsolidation()```

File: cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowNodeState.java [hunk 2]
```diff
@@ -60,6 +80,14 @@ public NodeStatus getNodeStatus() {
     return nodeStatus;
   }
 
+  /**
+   * Return the field operations to be recorded for this node. Field operations will be
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowNodeState.java:84]: Isn't it "recorded by this node"?

File: cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java [hunk 3]
```diff
@@ -66,4 +67,14 @@
    */
   @Beta
   ProgramState getState();
+
+  /**
+   * Call this method to disable emitting the field lineage operations from nodes running inside Workflow.
+   * Disabling should happen from {@link AbstractWorkflow#initialize} method at which point no node
+   * has been executed yet. Calling this method means Workflow is taking responsibility of emitting
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java:74]: How does the workflow do this? In its destroy() method? Or in a custom action? Does an action have the necessary APIs?

File: cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java [hunk 4]
```diff
@@ -66,4 +67,14 @@
...
+   * the field operation. Field operations will be available to workflow as {@link WorkflowNodeState}
+   * by calling {@link #getNodeStates} method.If workflow does not call {@link LineageRecorder#record}
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java:76]: missing space after period

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceLifecycleContext.java [hunk 5]
```diff
@@ -324,4 +326,10 @@ public MessageFetcher getMessageFetcher() {
     // TODO: CDAP-7807
     throw new UnsupportedOperationException("Messaging is not supported in MapReduce task-level context");
   }
+
+  @Override
+  public void record(Collection<? extends Operation> operations) {
+    throw new UnsupportedOperationException("Recording field lineage operations is not supported in MapReduce " +
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceLifecycleContext.java:332]: is this task-level? It is the MRLifecycleContext, not the MRTaskLifecycleContext

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java [hunk 6]
```diff
@@ -601,6 +611,23 @@ private void destroy() {
...
+
+    // MapReduce program ran as a part of Workflow
+    try {
+      Set<Operation> fieldLineageOperations = context.getFieldLineageOperations();
+      if (!workflowInfo.isDisableFieldLineageEmitFromNodes() && !fieldLineageOperations.isEmpty()) {
+        FieldLineageInfo info = new FieldLineageInfo(fieldLineageOperations);
+        fieldLineageWriter.write(mapReduceRunId, info);
+      }
+    } catch (Throwable t) {
+      LOG.warn("Failed to store the field lineage operations for MapReduce {}", mapReduceRunId, t);
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java:629]: is it store? Or emit?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java [hunk 7]
```diff
@@ -401,7 +410,6 @@ private void executeCustomAction(final WorkflowActionNode node, InstantiatorFact
                                                                     pluginInstantiator, secureStore,
                                                                     secureStoreManager, messagingService);
     customActionExecutor = new CustomActionExecutor(context, instantiator, classLoader);
-    status.put(node.getNodeId(), node);
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java:404]: Why this change? and the one below?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowProgramInfo.java [hunk 8]
```diff
@@ -53,23 +54,25 @@ public static WorkflowProgramInfo create(Arguments arguments) {
     String workflowRunId = arguments.getOption(ProgramOptionConstants.WORKFLOW_RUN_ID);
     String programNameInWorkflow = arguments.getOption(ProgramOptionConstants.PROGRAM_NAME_IN_WORKFLOW);
     String workflowToken = arguments.getOption(ProgramOptionConstants.WORKFLOW_TOKEN);
-
+    boolean disableFieldLineageEmitFromNodes = Boolean.parseBoolean(ProgramOptionConstants.DISABLE_FIELD_LINEAGE_EMIT);
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowProgramInfo.java:57]: that would be always false? Because that String is "disableFieldLineageEmit". Don't you want to parse arguments.getOption(ProgramOptionConstants.DISABLE_FIELD_LINEAGE_EMIT)? and only if it is not null?

**Review by anew — COMMENTED** [review]
See my last comment - this bug would always make it false. How have you tested it? Do you have unit tests? Also, we need an example that demonstrates how to use this feature.

**Comment by sagarkapare** [comment]
closing in favor of #10231
